src/main/java/uk/gov/companieshouse/api/accounts/service/impl/CurrentPeriodServiceImpl.java

```diff
@@ -1,33 +1,135 @@
+    }
+
+    @Override
+    public ResponseObject<CurrentPeriod> create(CurrentPeriod currentPeriod,
+            Transaction transaction,
+            String companyAccountId, String requestId)
+            throws DataException {
+
+        String selfLink = createSelfLink(transaction, companyAccountId);
+        initLinks(currentPeriod, selfLink);
+        currentPeriod.setEtag(GenerateEtagUtil.generateEtag());
+        currentPeriod.setKind(Kind.CURRENT_PERIOD.getValue());
+        CurrentPeriodEntity currentPeriodEntity = currentPeriodTransformer.transform(currentPeriod);
+
+        final Map<String, Object> debugMap = new HashMap<>();
+        debugMap.put("transaction_id", transaction.getId());
+        debugMap.put("company_accounts_id", companyAccountId);
+
+        try {
+            currentPeriodEntity.setId(generateID(companyAccountId));
+            currentPeriodRepository.insert(currentPeriodEntity);
+            smallFullService.addLink(companyAccountId, LinkType.SMALL_FULL, selfLink);
+        } catch (DuplicateKeyException dke) {
+            LOGGER.errorContext(requestId, dke, debugMap);
+            return new ResponseObject<>(ResponseStatus.DUPLICATE_KEY_ERROR, null);
+        } catch (MongoException me) {
+            DataException dataException = new DataException(
+                    "Failed to insert " + ResourceName.SMALL_FULL.getName(), me);
+            LOGGER.errorContext(requestId, dataException, debugMap);
+            throw dataException;
+        }
+
+        return new ResponseObject<>(ResponseStatus.CREATED, currentPeriod);
     }
 
     @Override
-    public void addKind(CurrentPeriod rest) {
-        rest.setKind(Kind.CURRENT_PERIOD.getValue());
+    public ResponseObject<CurrentPeriod> findById(String id) {
+        CurrentPeriodEntity currentPeriodEntity = currentPeriodRepository.findById(id).orElse(null);
+        if (currentPeriodEntity == null) {
+            return new ResponseObject<>(ResponseStatus.NOT_FOUND);
+        }
+        CurrentPeriod currentPeriod = currentPeriodTransformer.transform(currentPeriodEntity);
+        return new ResponseObject<>(ResponseStatus.FOUND, currentPeriod);
     }
 
     @Override
-    public String getResourceName() {
-        return "current-period";
+    public String generateID(String value) {
+        String unencryptedId = value + "-" + ResourceName.CURRENT_PERIOD.getName();
+        byte[] id = messageDigest.digest(
+                unencryptedId.getBytes(StandardCharsets.UTF_8));
+        return Base64.getUrlEncoder().encodeToString(id);
+    }
+
+    public void initLinks(CurrentPeriod currentPeriod, String link) {
+        Map<String, String> map = new HashMap<>();
+        map.put(LinkType.SELF.getLink(), link);
+        currentPeriod.setLinks(map);
+    }
+
+    @Override
+    public void addLink(String id, LinkType linkType, String link) {
+        CurrentPeriodEntity currentPeriodEntity = currentPeriodRepository.findById(id)
```

Comment:
as above - findById should be throwing DataException which can be passed up the chain